metrics/cuad/cuad.py

```diff
@@ -0,0 +1,117 @@
+      year={2021}
+}
+"""
+
+_DESCRIPTION = """
+This metric wrap the official scoring script for version 1 of the Contract
+Understanding Atticus Dataset (CUAD).
+
+Contract Understanding Atticus Dataset (CUAD) v1 is a corpus of more than 13,000 labels in 510
+commercial legal contracts that have been manually labeled to identify 41 categories of important
+clauses that lawyers look for when reviewing contracts in connection with corporate transactions.
+"""
+
+_KWARGS_DESCRIPTION = """
+Computes CUAD scores (EM, F1, AUPR, Precision@80%Recall, and Precision@90%Recall).
+Args:
+    predictions: List of question-answers dictionaries with the following key-values:
+        - 'id': id of the question-answer pair as given in the references (see below)
+        - 'prediction_text': list of possible texts for the answer, as a list of strings
+    references: List of question-answers dictionaries with the following key-values:
+        - 'id': id of the question-answer pair (see above),
+        - 'answers': a Dict in the CUAD dataset format
+            {
+                'text': list of possible texts for the answer, as a list of strings
+                'answer_start': list of start positions for the answer, as a list of ints
+            }
+            Note that answer_start values are not taken into account to compute the metric.
+Returns:
+    'exact_match': Exact match (the normalized answer exactly match the gold answer)
+    'f1': The F-score of predicted tokens versus the gold answer
+    'aupr': Area Under the Precision-Recall curve
+    'prec_at_80_recall': Precision at 80% recall
+    'prec_at_90_recall': Precision at 90% recall
+Examples:
+
+    >>> predictions = [{'prediction_text': ['The seller:', 'The buyer/End-User: Shenzhen LOHAS Supply Chain Management Co., Ltd.'], 'id': 'LohaCompanyltd_20191209_F-1_EX-10.16_11917878_EX-10.16_Supply Agreement__Parties'}]
+    >>> references = [{'answers': {'answer_start': [143, 49], 'text': ['The seller:', 'The buyer/End-User: Shenzhen LOHAS Supply Chain Management Co., Ltd.']}, 'id': 'LohaCompanyltd_20191209_F-1_EX-10.16_11917878_EX-10.16_Supply Agreement__Parties'}]
+    >>> cuad_metric = datasets.load_metric("cuad")
+    >>> results = cuad_metric.compute(predictions=predictions, references=references)
+    >>> print(results)
+    {'exact_match': 100.0, 'f1': 100.0, 'aupr': 0.0, 'prec_at_80_recall': 1.0, 'prec_at_90_recall': 1.0}
+"""
+
+
+@datasets.utils.file_utils.add_start_docstrings(_DESCRIPTION, _KWARGS_DESCRIPTION)
+class CUAD(datasets.Metric):
+    def _info(self):
+        return datasets.MetricInfo(
+            description=_DESCRIPTION,
+            citation=_CITATION,
+            inputs_description=_KWARGS_DESCRIPTION,
+            features=datasets.Features(
+                {
+                    "predictions": {
+                        "id": datasets.Value("string"),
+                        "prediction_text": datasets.features.Sequence(datasets.Value("string")),
```

Comment:
```suggestion
                        "prediction_text": datasets.Value("string"),
```
There should only be one prediction per question right ? Or is this something specific to CUAD ?

Reply:
I looked at the test samples in the dataset and found that few examples have multiple answers. In order to accomodate those answers too I changed it to `Sequence`

Reply:
Should I add a comment in the code too? In the docs though I've mentioned that 

> predictions: List of question-answers dictionaries with the following key-values:
>         - 'id': id of the question-answer pair as given in the references (see below)
>         - 'prediction_text': list of possible texts for the answer, as a list of strings

Reply:
> I looked at the test samples in the dataset and found that few examples have multiple answers. In order to accomodate those answers too I changed it to Sequence

If ome examples have multiple answers then it means that the `references` feature should allow for multiple reference answers (which is currently the case since you used Sequence in `references`). Not sure why you'd need a Sequence for `prediction_text` ?

Reply:
We'll need all possible predictions to calculate precision and recall. Say if I use only one prediction_text when my references are multiple then it'll increase my `false_negatives` and I'll get low value of `recall` according to the evaluation script

Reply:
Usually when someone evaluates a model, there's only one model prediction per question right ? Is this different for CUAD ?

Reply:
One can get multiple or single prediction depending on `conf` score [here (official evaluation script)](https://github.com/TheAtticusProject/cuad/blob/4e6b4f3a938dbc0498f800cb50046fe5f207bfcc/evaluate.py#L27). This is how they are building the `prediction_text` to evaluate precision and recall.

Reply:
Ok that makes sense !! Maybe mention in the docstring that users can use multiple answers depending on a threshold on the confidence probability of each prediction ?